README.md
# LinuxMSSQL-Toolbox

A collection of scripts to aid working with SQL Server on Linux or even Linux in general based on my personal field experience.
Health check script for a Linux server, includes SQL and Pacemaker checks

Scripts are all provided "as-is" with no waranty.

Tested on the following versions of Linux:

- RHEL

## Scripts folder

This is a general collection of scripts including

| Script Name | Purpose |
| backup_cleanup.sh | A script to clean up SQL Server backup locations as part of regular maintenance |

## HealthCheck folder

This is a script used to check the health of a server. Key features include:

- CPU, Disk, Memory checks
- SQL Server checks
- Pacemaker checks

## Change Log

See the [Change Log](./CHANGELOG.md) for the full changes.

## License

This project is released under the [MIT License](./LICENSE)

## Contributors

- Matticusau [GitHub](https://github.com/Matticusau) | [twitter](https://twitter.com/matticusau)


CHANGELOG.md
# Change Log

All notable changes to the project will be documented in this file. Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- None

## [0.1.4]

### Added

- None

### Changed

- HealthCheck now includes uptime info
- Added NOCOUNT to sql scripts where it was missing

## [0.1.3]

### Added

- Initial release
- HealthCheck
- backup_cleanup.sh

### Changed

- None


Scripts/archive_logs.sh
#!/bin/bash
# Purpose:  Archives log files based on parameters
# Author:   MLavery (PFE)
# Date:     22/06/2018
# Version:  0.1.0
#
# Version   When          Who       What
# 0.1.0     22/06/2018    MLavery   Initial file
#
# make sure you have execute rights
# chmod +rwx archive_logs.sh
#
# schedule with a cron job
# run "crontab -e" while logged in as the user with execute permissions on the script
# use vi to add this line to schedule every 15 mins
# 45 23 * * * /home/<user>/archive_logs.sh > /dev/null 2>&1
#

NOW=$(date +%Y_%m_%d_%H_%M_%S)
LOG_FILE_DIR=/home/<user>/pfecheck/log
LOG_FILE_RETENTION=1;

# set the working directory
cd $LOG_FILE_DIR;

# create the temporary folder
mkdir $NOW;

# move the log files to the temporary folder
find $LOG_FILE_DIR/* -type f -mtime +$LOG_FILE_RETENTION -exec mv {} ./$NOW \;

# compress the folder
tar cfj $NOW.bz2 -C $NOW .

# remove the temporary folder
rm -r ./$NOW


Scripts/backup_cleanup.sh
#!/bin/bash
# Purpose:  Cleans up the MSSQL backups
#           Part of the LinuxMSSQL-Toolbox - https://github.com/Matticusau/LinuxMSSQL-Toolbox
# Author:   Matticusau / MLavery
# Date:     19/06/2018
# Version:  0.1.0
#
# Version   When          Who       What
# 0.1.0     20/06/2018    MLavery   Initial version
# 
# make sure you have execute rights
# chmod +rwx backup_cleanup.sh
#
# schedule with a cron job
# run "sudo crontab -u root -e" while logged in as the user with execute permissions on the script
# use vi to add this line to schedule every 15 mins
# 0 23 * * * /backup/backup_cleanup.sh > /dev/null 2>&1
#

# settings
FULL_BKP_DIR=/backup/full
DIFF_BKP_DIR=/backup/diff
TLOG_BKP_DIR=/backup/log
FULL_BKP_RETENTION=14;
DIFF_BKP_RETENTION=14;
TLOG_BKP_RETENTION=1;

# current time
NOW=$(date +%Y_%m_%d_%H_%M_%S)

echo -e "Clean up starting $(date +%Y_%m_%d_%H_%M_%S)";

# clean up full backups
echo -e "Start cleanup process >> $NOW" >> $FULL_BKP_DIR/cleanup.log;
CMD="find $FULL_BKP_DIR/*.bak -type f -mtime +$FULL_BKP_RETENTION"
COUNT=$($CMD -printf '.' 2>/dev/null | wc -c);
echo -e ">>> $COUNT Diff backups";
$($CMD -exec ls -lrt {} \; 2>/dev/null) >> $FULL_BKP_DIR/cleanup.log
sudo $CMD -exec rm -f {} \;

# clean up differential backups
echo -e "Start cleanup process >> $NOW" >> $DIFF_BKP_DIR/cleanup.log;
CMD="find $DIFF_BKP_DIR/*.bak -type f -mtime +$DIFF_BKP_RETENTION"
COUNT=$($CMD -printf '.' 2>/dev/null | wc -c);
echo -e ">>> $COUNT Diff backups";
$($CMD -exec ls -lrt {} \; 2>/dev/null) >> $DIFF_BKP_DIR/cleanup.log
sudo $CMD -exec rm -f {} \;

# clean up transaction log backups
echo -e "Start cleanup process >> $NOW" >> $TLOG_BKP_DIR/cleanup.log;
CMD="find $TLOG_BKP_DIR/*.trn -type f -mtime +$TLOG_BKP_RETENTION"
COUNT=$($CMD -printf '.' 2>/dev/null | wc -c);
echo -e ">>> $COUNT Transaction Log backups";
$($CMD -exec ls -lrt {} \; 2>/dev/null) >> $TLOG_BKP_DIR/cleanup.log
sudo $CMD -exec rm -f {} \;

echo -e "Clean up finished $(date +%Y_%m_%d_%H_%M_%S)";


HealthCheck/pfe_check.sh
#!/bin/bash
# Purpose:  Provides a check of the Linuxserver for use with monitoring / health check activities
#           Part of the LinuxMSSQL-Toolbox - https://github.com/Matticusau/LinuxMSSQL-Toolbox
# Author:   Matticusau / MLavery (PFE)
# Date:     19/06/2018
# Version:  0.1.4
#
# Version   When          Who       What
# 0.1.1     20/06/2018    MLavery   Added log file clean up, SQL scripts, work file logging
# 0.1.2     21/06/2018    MLavery   Added Top Waits check
# 0.1.3     22/06/2018    MLavery   Added VLFs, DB Space
# 0.1.4     04/07/2018    MLavery   Added uptime info
#
# make sure you have execute rights
# chmod +rwx pfe_check.sh
#
# make sure sqlcmd (mssql-tools) is installed and the bin folder is in the path variable
# PATH=$PATH:/opt/mssql-tools/bin
#
# schedule with a cron job
# run "crontab -e" while logged in as the user with execute permissions on the script
# use vi to add this line to schedule every 15 mins
# */30 * * jun * /home/<user>/pfecheck/pfe_check.sh > /dev/null 2>&1
#

NOW=$(date +%Y_%m_%d_%H_%M_%S)
SCRIPT_FILE_DIR=/home/<user>/pfecheck
LOG_FILE_DIR=/home/<user>/pfecheck/log
LOG_FILE="$LOG_FILE_DIR/$(hostname)_$NOW.log";
LOG_FILE_RETENTION=7;
DISK_MNT_FILTER="/log|/data|/backup|/var";
SQL_LOG_LINES=25;
SQL_USER=
SQL_PWD=
CHECK_SQL=1;
CHECK_SQL_AG=1;
CHECK_PACEMAKER=1;

# set the working directory
cd $LOG_FILE_DIR;

# clean up an old log files
COUNT=$(find $LOG_FILE_DIR/*.log -mtime +$LOG_FILE_RETENTION | wc -c);
if [ $COUNT -gt 0 ];
then
    echo "Cleaning up $COUNT old log files...";
    find "$LOG_FILE_DIR/*.log" -mtime +$LOG_FILE_RETENTION -exec rm -f {} \; 
fi

# set the work file name
WORK_FILE="$LOG_FILE.work"

# initialise the file each time
echo -e "" > $WORK_FILE;

# header
echo -e "**************************************************************\nPFE Check" >> $WORK_FILE;
echo -e $(hostname) >> $WORK_FILE;
echo -e $NOW >> $WORK_FILE;
echo -e "**************************************************************" >> $WORK_FILE;

# uptime
echo -e "\n-------------------------------\nUp Time\n-------------------------------" >> $WORK_FILE;
echo -e "$(hostname): $(awk '{ print $1/60/60/24 " days" }' /proc/uptime)" >> $WORK_FILE;

# cpu usage
echo -e "\n-------------------------------\nCPU (top)\n-------------------------------" >> $WORK_FILE;
top -b -n1 | head -n 20 >> $WORK_FILE;

# cpu usage
echo -e "\n-------------------------------\nCPU IO (iostat)\n-------------------------------" >> $WORK_FILE;
iostat >> $WORK_FILE;

# disk IO
echo -e "\n-------------------------------\nDisk IO\n-------------------------------" >> $WORK_FILE;
dstat -tdD total --output $WORK_FILE 30 5

# disk space
echo -e "\n-------------------------------\nDisk Space\n-------------------------------" >> $WORK_FILE;
echo -e "Date\tSize\tUsed\tAvail\tUsePct\tMnt" >> $WORK_FILE;
df -h | grep -E "($DISK_MNT_FILTER)" | awk -v out_file="$WORK_FILE" ' BEGIN { OFS="\t"; ORS="\n" } { print strftime("[%Y-%m-%d %H:%M:%S]"), $2, $3, $4,  $5, $6 >> out_file}'
# df -h | awk -v out_file="$WORK_FILE" ' BEGIN { OFS="\t"; ORS="\n" } { print strftime("[%Y-%m-%d %H:%M:%S]"), $2, $3, $4,  $5, $6 >> out_file}'

# SQL checks
if [ $CHECK_SQL -eq 1 ];
then
    # SQL ErrorLog
    echo -e "\n-------------------------------\nSQL ERRORLOG ($SQL_LOG_LINES lines)\n-------------------------------" >> $WORK_FILE;
    sudo tail -n $SQL_LOG_LINES /var/opt/mssql/log/errorlog  >> $WORK_FILE;

    # SQL ErrorLog Errors
    echo -e "\n-------------------------------\nSQL ERRORLOG Errors\n-------------------------------" >> $WORK_FILE;
    sudo cat /var/opt/mssql/log/errorlog | grep error >> $WORK_FILE;

    # VLFs
    echo -e "\n-------------------------------\nUser DB File Free Space\n-------------------------------" >> $WORK_FILE;
    /opt/mssql-tools/bin/sqlcmd -S. -U$SQL_USER -P$SQL_PWD -i $SCRIPT_FILE_DIR/MonitorDBSpace.sql -W -s "," >> $WORK_FILE;

    # SQL ErrorLog Errors
    echo -e "\n-------------------------------\nLog File Reuse\n-------------------------------" >> $WORK_FILE;
    /opt/mssql-tools/bin/sqlcmd -S. -U$SQL_USER -P$SQL_PWD -Q "SELECT name, log_reuse_wait_desc FROM sys.databases ORDER BY name" -W -s "," >> $WORK_FILE;

    # SQL Top Waits
    echo -e "\n-------------------------------\nTop Waits\n-------------------------------" >> $WORK_FILE;
    /opt/mssql-tools/bin/sqlcmd -S. -U$SQL_USER -P$SQL_PWD -i $SCRIPT_FILE_DIR/MonitorTopWaits.sql -W -s "," >> $WORK_FILE;

    # VLFs
    echo -e "\n-------------------------------\nVLFs (>1000)\n-------------------------------" >> $WORK_FILE;
    /opt/mssql-tools/bin/sqlcmd -S. -U$SQL_USER -P$SQL_PWD -i $SCRIPT_FILE_DIR/MonitorVLFs.sql -W -s "," >> $WORK_FILE;

fi

# Availability Group checks
if [ $CHECK_SQL_AG -eq 1 ];
then
    # AG Replica Roles
    echo -e "\n-------------------------------\nAG Replica Roles\n-------------------------------" >> $WORK_FILE;
    /opt/mssql-tools/bin/sqlcmd -S. -U$SQL_USER -P$SQL_PWD -i $SCRIPT_FILE_DIR/MonitorAGReplicaRoles.sql -W -s "," >> $WORK_FILE;

    # AG Health
    echo -e "\n-------------------------------\nAG Replica Health\n-------------------------------" >> $WORK_FILE;
    /opt/mssql-tools/bin/sqlcmd -S. -U$SQL_USER -P$SQL_PWD -i $SCRIPT_FILE_DIR/MonitorAGReplicaHealth.sql -W -s "," >> $WORK_FILE;

     # AG Database Health
    echo -e "\n-------------------------------\nAG Database Health\n-------------------------------" >> $WORK_FILE;
    /opt/mssql-tools/bin/sqlcmd -S. -U$SQL_USER -P$SQL_PWD -i $SCRIPT_FILE_DIR/MonitorAGDbHealth.sql -W -s "," >> $WORK_FILE;

fi

# Pacemaker checks
if [ $CHECK_PACEMAKER -eq 1 ];
then
    # PCS status
    echo -e "\n-------------------------------\nPCS Status\n-------------------------------" >> $WORK_FILE;
    sudo pcs status >> $WORK_FILE;

    # PCS Location Constraints 
    echo -e "\n-------------------------------\nPCS Location Constraints\n-------------------------------" >> $WORK_FILE;
    sudo pcs constraint list --full >> $WORK_FILE;
fi

# rename the work file to the log file
mv $WORK_FILE $LOG_FILE;

# footer
# echo -e "\n**************************************************************\nPFE Monitor Complete\n**************************************************************\n\n\n" >> $WORK_FILE;

